// v3-dragger/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useDesktopDragger = exports.useMobileDragger = exports.useDragger = exports.vDraggerUnMount = exports.vDraggerBeforeMount = exports.isTouchEvent = void 0;
const vue_1 = require("vue");
const isTouchEvent = (e) => !!e.touches;
exports.isTouchEvent = isTouchEvent;
const beforeEventHandler = (e) => {
    var _a;
    if ((0, exports.isTouchEvent)(e)) {
        return;
    }
    // get rid of the ghost image of the component
    const img = new Image();
    img.src =
        'data:image/gif;base64,R0lGODlhAQABAIAAAAUEBAAAACwAAAAAAQABAAACAkQBADs=';
    (_a = e.dataTransfer) === null || _a === void 0 ? void 0 : _a.setDragImage(img, 0, 0);
};
const dragStartGeneric = (e, dragYOffset, dragYStart, dragXOffset, dragXStart, clientY, clientX, customDragStart) => {
    if (customDragStart) {
        customDragStart(e, dragYOffset, dragYStart, dragXOffset, dragXStart, clientY, clientX);
        return;
    }
    else {
        dragYStart.value = clientY;
        dragXStart.value = clientX;
    }
};
const dragMoveGeneric = (e, dragYOffset, dragYStart, dragXOffset, dragXStart, clientY, clientX, customDragMove) => {
    /**
     * When letting go of the mouse button, clientY and clientX both become
     * 0 for some reason, at least in chromium-based browsers.
     */
    if (clientY === 0 && clientX === 0) {
        return;
    }
    else {
        if (customDragMove) {
            customDragMove(e, dragYOffset, dragYStart, dragXOffset, dragXStart, clientY, clientX);
            return;
        }
        else {
            dragYOffset.value = dragYStart.value - clientY;
            dragXOffset.value = dragXStart.value - clientX;
        }
    }
};
const dragEndGeneric = (e, dragYOffset, dragYStart, dragXOffset, dragXStart, clientY, clientX, customDragEnd) => {
    if (customDragEnd) {
        customDragEnd(e, dragYOffset, dragYStart, dragXOffset, dragXStart, clientY, clientX);
    }
    /**
     * By default, we don't do anything on end. This function
     * is declared here to get it's types down.
     */
};
const vDraggerBeforeMount = (el, binding) => {
    const { dragYOffset, dragXOffset, dragStart, dragMove } = (0, exports.useDragger)();
    // drag listeners
    el.addEventListener('touchstart', (e) => {
        var _a, _b;
        dragStart(e);
        (_b = (_a = binding.value) === null || _a === void 0 ? void 0 : _a.dragStart) === null || _b === void 0 ? void 0 : _b.call(_a, e, dragYOffset, dragXOffset);
    });
    el.addEventListener('dragstart', (e) => {
        var _a, _b;
        dragStart(e);
        (_b = (_a = binding.value) === null || _a === void 0 ? void 0 : _a.dragStart) === null || _b === void 0 ? void 0 : _b.call(_a, e, dragYOffset, dragXOffset);
    });
    el.addEventListener('touchmove', (e) => {
        var _a, _b;
        dragMove(e);
        (_b = (_a = binding.value) === null || _a === void 0 ? void 0 : _a.dragMove) === null || _b === void 0 ? void 0 : _b.call(_a, e, dragYOffset, dragXOffset);
    });
    el.addEventListener('drag', (e) => {
        var _a, _b;
        dragMove(e);
        (_b = (_a = binding.value) === null || _a === void 0 ? void 0 : _a.dragMove) === null || _b === void 0 ? void 0 : _b.call(_a, e, dragYOffset, dragXOffset);
    });
    el.addEventListener('touchend', (e) => {
        var _a, _b;
        (_b = (_a = binding.value) === null || _a === void 0 ? void 0 : _a.dragEnd) === null || _b === void 0 ? void 0 : _b.call(_a, e, dragYOffset, dragXOffset);
    });
    el.addEventListener('dragend', (e) => {
        var _a, _b;
        (_b = (_a = binding.value) === null || _a === void 0 ? void 0 : _a.dragEnd) === null || _b === void 0 ? void 0 : _b.call(_a, e, dragYOffset, dragXOffset);
    });
    // ux
    el.setAttribute('draggable', 'true');
    el.style.setProperty('cursor', 'grab');
    // initial set
    el.style.setProperty('--v-dragger-y-offset', `${-dragYOffset.value}`);
    el.style.setProperty('--v-dragger-x-offset', `${-dragXOffset.value}`);
    (0, vue_1.watch)([dragYOffset, dragXOffset], () => {
        el.style.setProperty('--v-dragger-y-offset', `${-dragYOffset.value}`);
        el.style.setProperty('--v-dragger-x-offset', `${-dragXOffset.value}`);
    });
};
exports.vDraggerBeforeMount = vDraggerBeforeMount;
const vDraggerUnMount = (el) => {
    /* eslint-disable @typescript-eslint/no-unused-vars */
    /* eslint-disable @typescript-eslint/no-empty-function */
    el.removeEventListener('touchstart', (e) => { });
    el.removeEventListener('dragstart', (e) => { });
    el.removeEventListener('touchmove', (e) => { });
    el.removeEventListener('drag', (e) => { });
    el.removeEventListener('touchend', (e) => { });
    el.removeEventListener('dragend', (e) => { });
    /* eslint-enable @typescript-eslint/no-unused-vars */
    /* eslint-enable @typescript-eslint/no-empty-function */
};
exports.vDraggerUnMount = vDraggerUnMount;
const useDragger = (customDragEndGeneric, customDragMoveGeneric, customDragStartGeneric) => {
    const dragYStart = (0, vue_1.ref)(0);
    const dragYOffset = (0, vue_1.ref)(0);
    const dragXStart = (0, vue_1.ref)(0);
    const dragXOffset = (0, vue_1.ref)(0);
    const dragStart = (e) => {
        beforeEventHandler(e);
        (0, exports.isTouchEvent)(e)
            ? dragStartGeneric(e, dragYOffset, dragYStart, dragXOffset, dragXStart, e.touches[0].clientY, e.touches[0].clientX, customDragStartGeneric)
            : dragStartGeneric(e, dragYOffset, dragYStart, dragXOffset, dragXStart, e.clientY, e.clientX, customDragStartGeneric);
    };
    const dragMove = (e) => {
        beforeEventHandler(e);
        (0, exports.isTouchEvent)(e)
            ? dragMoveGeneric(e, dragYOffset, dragYStart, dragXOffset, dragXStart, e.touches[0].clientY, e.touches[0].clientX, customDragMoveGeneric)
            : dragMoveGeneric(e, dragYOffset, dragYStart, dragXOffset, dragXStart, e.clientY, e.clientX, customDragMoveGeneric);
    };
    const dragEnd = (e) => {
        (0, exports.isTouchEvent)(e)
            ? dragEndGeneric(e, dragYOffset, dragYStart, dragXOffset, dragXStart, e.touches[0].clientY, e.touches[0].clientX, customDragEndGeneric)
            : dragEndGeneric(e, dragYOffset, dragYStart, dragXOffset, dragXStart, e.clientY, e.clientX, customDragEndGeneric);
    };
    const resetState = () => {
        dragYStart.value = 0;
        dragYOffset.value = 0;
        dragXStart.value = 0;
        dragXOffset.value = 0;
    };
    return {
        dragYStart,
        dragYOffset,
        dragXStart,
        dragXOffset,
        dragStart,
        dragMove,
        dragEnd,
        resetState,
    };
};
exports.useDragger = useDragger;
const useMobileDragger = (customDragEndGeneric, customDragMoveGeneric, customDragStartGeneric) => {
    const touchYStart = (0, vue_1.ref)(0);
    const touchYOffset = (0, vue_1.ref)(0);
    const touchXStart = (0, vue_1.ref)(0);
    const touchXOffset = (0, vue_1.ref)(0);
    const touchStart = (e) => {
        dragStartGeneric(e, touchYOffset, touchYStart, touchXOffset, touchXStart, e.touches[0].clientY, e.touches[0].clientX, customDragStartGeneric);
    };
    const touchMove = (e) => {
        dragMoveGeneric(e, touchYOffset, touchYStart, touchXOffset, touchXStart, e.touches[0].clientY, e.touches[0].clientX, customDragMoveGeneric);
    };
    const touchEnd = (e) => {
        dragEndGeneric(e, touchYOffset, touchYStart, touchXOffset, touchXStart, e.touches[0].clientY, e.touches[0].clientX, customDragEndGeneric);
    };
    const resetState = () => {
        touchYStart.value = 0;
        touchYOffset.value = 0;
        touchXStart.value = 0;
        touchXOffset.value = 0;
    };
    return {
        touchYStart,
        touchXStart,
        touchYOffset,
        touchXOffset,
        touchStart,
        touchMove,
        touchEnd,
        resetState,
    };
};
exports.useMobileDragger = useMobileDragger;
const useDesktopDragger = (customDragEndGeneric, customDragMoveGeneric, customDragStartGeneric) => {
    const dragYStart = (0, vue_1.ref)(0);
    const dragYOffset = (0, vue_1.ref)(0);
    const dragXStart = (0, vue_1.ref)(0);
    const dragXOffset = (0, vue_1.ref)(0);
    const dragStart = (e) => {
        beforeEventHandler(e);
        dragStartGeneric(e, dragYOffset, dragYStart, dragXOffset, dragXStart, e.clientY, e.clientX, customDragStartGeneric);
    };
    const dragMove = (e) => {
        beforeEventHandler(e);
        dragMoveGeneric(e, dragYOffset, dragYStart, dragXOffset, dragXStart, e.clientY, e.clientX, customDragMoveGeneric);
    };
    const dragEnd = (e) => {
        dragEndGeneric(e, dragYOffset, dragYStart, dragXOffset, dragXStart, e.clientY, e.clientX, customDragEndGeneric);
    };
    const resetState = () => {
        dragYStart.value = 0;
        dragYOffset.value = 0;
        dragXStart.value = 0;
        dragXOffset.value = 0;
    };
    return {
        dragYStart,
        dragXStart,
        dragYOffset,
        dragXOffset,
        dragStart,
        dragMove,
        dragEnd,
        resetState,
    };
};
exports.useDesktopDragger = useDesktopDragger;
